File: Codeforces/E/168/D.cpp
#include <bits/stdc++.h>
using namespace std;

#define INF (2e9+10)
#define int long long

// Optimized input
void fast_io() {
    ios::sync_with_stdio(false);
    cin.tie(0);
}

// Function to solve each test case
void solve() {
    int n;
    cin >> n;

    vector<int> A(n), P(n);
    for (int &x : A) cin >> x;
    for (int i = 1; i < n; ++i) cin >> P[i], P[i]--;

    vector<vector<int>> child(n);
    for (int i = 1; i < n; ++i) {
        child[P[i]].push_back(i);
    }

    vector<int> depth(n, 0);
    queue<int> q;
    q.push(0);

    while (!q.empty()) {
        int cur = q.front();
        q.pop();
        for (int o : child[cur]) {
            depth[o] = depth[cur] + 1;
            q.push(o);
        }
    }

    auto check = [&](int mid) -> bool {
        if (A[0] >= mid) return true;
        vector<int> B = A;

        queue<pair<int, int>> q;
        for (int o : child[0]) {
            q.emplace(o, mid - B[0]);
        }

        while (!q.empty()) {
            auto [cur, need] = q.front();
            q.pop();
            B[cur] -= need;
            if (B[cur] + INF < 0){
                return false;
            }

            if (child[cur].empty()) {
                if (B[cur] < 0) return false;
            }

            int new_need = B[cur] >= 0 ? 0 : -B[cur];
            for (int o : child[cur]) {
                q.emplace(o, need + new_need);
            }
        }
        return true;
    };

    int l = 0, r = INF;
    while (l < r) {
        int mid = (l + r) >> 1;
        if (check(mid)) {
            l = mid + 1;
        } else {
            r = mid;
        }
    }
    cout << l - 1 << '\n';
}

signed main() {
    fast_io();
    int t;
    cin >> t;
    while (t--) {
        solve();
    }
    return 0;
}

File: Codeforces/E/168/D.py
'''
Hala Madrid!
https://www.zhihu.com/people/li-dong-hao-78-74
'''

import sys
import os
from io import BytesIO, IOBase
BUFSIZE = 8192
class FastIO(IOBase):
    newlines = 0
    def __init__(self, file):
        self._fd = file.fileno()
        self.buffer = BytesIO()
        self.writable = "x" in file.mode or "r" not in file.mode
        self.write = self.buffer.write if self.writable else None
    def read(self):
        while True:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            if not b:
                break
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines = 0
        return self.buffer.read()
    def readline(self):
        while self.newlines == 0:
            b = os.read(self._fd, max(os.fstat(self._fd).st_size, BUFSIZE))
            self.newlines = b.count(b"\n") + (not b)
            ptr = self.buffer.tell()
            self.buffer.seek(0, 2), self.buffer.write(b), self.buffer.seek(ptr)
        self.newlines -= 1
        return self.buffer.readline()
    def flush(self):
        if self.writable:
            os.write(self._fd, self.buffer.getvalue())
            self.buffer.truncate(0), self.buffer.seek(0)
class IOWrapper(IOBase):
    def __init__(self, file):
        self.buffer = FastIO(file)
        self.flush = self.buffer.flush
        self.writable = self.buffer.writable
        self.write = lambda s: self.buffer.write(s.encode("ascii"))
        self.read = lambda: self.buffer.read().decode("ascii")
        self.readline = lambda: self.buffer.readline().decode("ascii")
sys.stdin, sys.stdout = IOWrapper(sys.stdin), IOWrapper(sys.stdout)
input = lambda: sys.stdin.readline().rstrip("\r\n")

def I():
    return input()
def II():
    return int(input())
def MI():
    return map(int, input().split())
def LI():
    return list(input().split())
def LII():
    return list(map(int, input().split()))
def GMI():
    return map(lambda x: int(x) - 1, input().split())

#------------------------------FastIO---------------------------------

from bisect import *
from heapq import *
from collections import *
from functools import *
from itertools import *
from time import *
from random import *
from math import log, gcd, sqrt, ceil

from types import GeneratorType
def bootstrap(f, stack=[]):
    def wrappedfunc(*args, **kwargs):
        if stack:
            return f(*args, **kwargs)
        else:
            to = f(*args, **kwargs)
            while True:
                if type(to) is GeneratorType:
                    stack.append(to)
                    to = next(to)
                else:
                    stack.pop()
                    if not stack:
                        break
                    to = stack[-1].send(to)
            return to
    return wrappedfunc

# seed(19981220)
# RANDOM = getrandbits(64)
 
# class Wrapper(int):
#     def __init__(self, x):
#         int.__init__(x)

#     def __hash__(self):
#         return super(Wrapper, self).__hash__() ^ RANDOM

# def TIME(f):

#     def wrap(*args, **kwargs):
#         s = perf_counter()
#         ret = f(*args, **kwargs)
#         e = perf_counter()

#         print(e - s, 'sec')
#         return ret
    
#     return wrap

inf = float('inf')

fmin = lambda x, y: x if x < y else y
fmax = lambda x, y: x if x > y else y

# @TIME
def solve(testcase):
    n = II()
    A = LII()
    P = LII()
    child = [[] for _ in range(n)]
    
    for i, v in enumerate(P, 1):
        child[v - 1].append(i)
        
    
    def f(u, v):
        return v * 200010 + u
    
    # depth = [0 for _ in range(n)]
    # q = deque()
    # q.append(0)
    
    # while q:
    #     cur = q.popleft()
    #     for o in child[cur]:
    #         depth[o] = depth[cur] + 1
    #         q.append(o)
    
    # print('check depth', depth)
    
    def check(mid):
        # if A[0] >= mid:
        #     return True
        
        B = A[:]
        
        q = deque()
        for o in child[0]:
            q.append(f(o, mid - B[0]))
        
        while q:
            cur = q.popleft()
            need, cur = divmod(cur, 200010)
            B[cur] -= need
            if B[cur] < -(10 ** 9 + 10):
                return False
            if not child[cur]:
                if B[cur] < 0:
                    # q.clear()
                    return False
                
            newneed = 0 if B[cur] >= 0 else -B[cur]
            for o in child[cur]:
                q.append(f(o, need + newneed))
        
        # q.clear()
        return True
    
    l, r = A[0], 2 * 10 ** 9 + 10
    while l < r:
        mid = l + r >> 1
        if check(mid):
            l = mid + 1
        else:
            r = mid
    
    print(l - 1)    

for testcase in range(II()):
    solve(testcase)